File: src/app/camera.py
import os
import time
import threading
from typing import Optional

import cv2
import numpy as np

from .stats import StreamStats

class CameraStream:
    """
    Захватывает кадры с камеры в отдельном потоке.
    Хранит только последний кадр.
    """
    def __init__(self, device:str, width: int = 854, height:int = 480, fps: int = 10):
        self.device = device
        self.width = width
        self.height = height
        self.fps = fps

        self._cap: Optional[cv2.VideoCapture] = None
        self._thread: Optional[threading.Thread] = None
        self._stop = threading.Event()

        self._lock = threading.Lock()
        self._last_frame: Optional[np.ndarray] = None

        self.stats = StreamStats()
    
    def start(self) -> None:
        if self._thread and self._thread.is_alive():
            return
        
        self._init_cap()

        self._stop.clear()
        self._thread = threading.Thread(target=self._loop, daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        if self._thread:
            self._thread.join(timeout=2.0)
        if self._cap:
            self._cap.release()
        
        self._cap = None
        self._thread = None

    def get_last_frame(self) -> Optional[np.ndarray]:
        with self._lock:
            return None if self._last_frame is None else self._last_frame.copy()
        
    def _loop(self) -> None:
        assert self._cap is not None
        while not self._stop.is_set():
            ret, frame = self._cap.read()
            if not ret or frame is None:
                self.stats.frames_drop += 1
                time.sleep(0.01)
                continue

            with self._lock:
                self._last_frame = frame
                self.stats.frames_ok += 1
                self.stats.last_frame_ts = time.time()

    def _init_cap(self) -> None:
        self._cap = cv2.VideoCapture(self.device, cv2.CAP_V4L2)

        self._cap.set(cv2.CAP_PROP_FRAME_WIDTH, self.width)
        self._cap.set(cv2.CAP_PROP_FRAME_HEIGHT, self.height)
        self._cap.set(cv2.CAP_PROP_FPS, self.fps)

def get_device_from_env() -> str:
    return os.environ.get("VIDEO_DEVICE", "dev/video2")

File: src/camera.cpp
#include "camera.h"
#include <sstream>

Camera::Camera(int index) : index_(index) {}

void Camera::request(int width, int height, double fps) {
    req_w_ = width;
    req_h_ = height;
    req_fps_ = fps;
}

bool Camera::open() {
    if (cap_.isOpened()) {
        cap_.release();
    }

    cap_.open(index_, cv::CAP_V4L2);
    if (!cap_.isOpened()) {
        cap_.open(index_, cv::CAP_ANY);
    }
    if (!cap_.isOpened()) return false;

    cap_.set(cv::CAP_PROP_FRAME_WIDTH,  req_w_);
    cap_.set(cv::CAP_PROP_FRAME_HEIGHT, req_h_);
    cap_.set(cv::CAP_PROP_FPS,          req_fps_);

    return true;
}

bool Camera::read(cv::Mat& outFrame) {
    if (!cap_.isOpened()) {
        return false;
    }
    return cap_.read(outFrame);
}

std::string Camera::info() const {
    std::ostringstream ss;
    if (!cap_.isOpened()) {
        ss << "Camera: not opened";
        return ss.str();
    }

    const double w   = cap_.get(cv::CAP_PROP_FRAME_WIDTH);
    const double h   = cap_.get(cv::CAP_PROP_FRAME_HEIGHT);
    const double fps = cap_.get(cv::CAP_PROP_FPS);

    ss << "Camera Info: Index: " << index_
       << ", Width: " << (int)w
       << ", Height: " << (int)h
       << ", FPS: " << fps;

    return ss.str();
}
